Standardize AST Constructors

markwpearce opened the pull request:
Standardizes all AstNodes such that they:

- take a single `options` object in the constructor
- constructor options keys end in `token` if it is a token or an identifier
- all tokens are accessible via a `.tokens` member on the node
- optional and/or mandatory keyword tokens do not need to be included in the constructor
- transpiles will just fill in a missing keyword if not included.

Addresses: #1007

TwitchBronBron commented on src/util.ts:
```diff
@@ -596,15 +597,30 @@ export class Util {
         }
 
-        if (position.line < range.start.line || (position.line === range.start.line && position.character < range.start.character)) {
+        if (this.comparePosition(position, range.start) < 0) {
+            return -1;
+        }
+        if (this.comparePosition(position, range.end) > 0) {
+            return 1;
+        }
+        return 0;
+    }
+
+    public comparePosition(a: Position | undefined, b: Position) {
+        //stop if the either position is missing
+        if (!a || !b) {
+            return 0;
+        }
+
+        if (a.line < b.line || (a.line === b.line && a.character < b.character)) {
             return -1;
         }
```
I feel like I had inlined this stuff due to optimization/performance reasons. but I don't remember that for sure, so I'm just adding a comment here to call it out in case we are looking for places to optimize again in the future.

TwitchBronBron reviewed (changes requested):
Hey, this looks pretty good. I've been starting at all the `xyzToken` names on the options, and I just really don't prefer the trailing `token` name. 

I think I'd prefer to remove the `Token` suffix from all the options props. TypeScript is already there to help us detect when an incorrect interface type is assigned, and provides hover/intellisense with what the props are. 

before:
```typescript
new AAMemberExpression({
    keyToken: createToken('awer'),
    colonToken: createToken('asdf'),
    value: VariableExpression('asdf'),
    commaToken: createToken('asdf')
})
```

after:
```
new AAMemberExpression({
    key: createToken('awer'),
    colon: createToken('asdf'),
    value: VariableExpression('asdf'),
    comma: createToken('asdf')
})
```

markwpearce commented on src/parser/Statement.ts:
```diff
@@ -125,30 +129,43 @@ export class Body extends Statement implements TypedefProvider {
+        equalsToken?: Token;
+        nameToken: Identifier;
+        value: Expression;
+        // TODO: add explicit typing to assignments
```
I went ahead and added it:


```
sub foo(y)
  x as integer = y ' x will be considered an integer by type system
end sub
```

```
sub foo2(node as roSGNode)
  x as integer = node.id  ' validation error - roSgNode.id is a string
end sub
```

TwitchBronBron replied:
Love this!

markwpearce commented:
I updated this PR from the comments, and from a few notes in https://github.com/rokucommunity/brighterscript/issues/1044

- Removed `Token` and `Keyword` suffixes for AST node constructor options and `tokens` member.
- all `=` tokens are called `equals` for consistency
- Added type declaration on Assignments (in Brighterscript)

```
sub foo(y)
  x as integer = y ' x will be considered an integer by type system
end sub

sub foo2(node as roSGNode)
  x as integer = node.id  ' validation error - roSgNode.id is a string
end sub
```

TwitchBronBron reviewed (approved):
Looks great, thanks!
